### PR TITLE
Reuse userAccount data for profile if possible.

### PR DESCRIPTION
**Pull request by davidjoy**
If we're looking at the profile of the authenticated user, skip calling fetchAccount a second time to load profile data and just re-use the user account data we loaded as part of frontend-auth's work.

It may be useful to review this by commit.  The first commit removes some data massaging we were doing that was ultimately getting in the way.  Before this PR, we were massaging data for the profile to change "levelOfEducation" to "education" as a nicety, but now we want to reuse userAccount data as the profile, that conversion is kind of a pain/was just going to make the process more complicated for little benefit.  So I removed it.  

The second commit relies on a change to frontend-auth to put a `loaded` flag in the userAccount part of the redux store state.  We use this to guarantee that the userAccount is loaded before we decide whether or not we should fetch the profile.  Technically for other users' profiles this makes our requests more serial, but I couldn't find a way to keep them fully parallel without a more significant refactor.

### CODE REVIEW

File: src/components/App.jsx [hunk 1]
```diff
@@ -24,41 +25,51 @@ class App extends Component {
     this.props.fetchUserAccount(userAccountApiService, username);
   }
 
+  renderContent() {
+    if (!this.props.loaded) {
```
Comment (davidjoy) [inline src/components/App.jsx]: Pulling this out into a separate function and conditionally returning on `loaded` allows us to defer calling `fetchProfile` in ConnectedProfilePage until we've gotten the userAccount data.  This is the serialization I was talking about in the PR description.  We don't even want to show the ConnectedProfilePage if the user isn't authenticated, so I didn't want to bother making requests for the profile data here in componentDidMount() before we know if we even need to.

File: src/sagas/RootSaga.js [hunk 2]
```diff
@@ -38,29 +38,39 @@ import { keepKeys } from '../services/utils';
 
 export function* handleFetchProfile(action) {
   const { username } = action.payload;
-  const currentUsername = yield select(state => state.authentication.username); // eslint-disable-line
+  const { isCurrentUserProfile, userAccount } = yield select(handleFetchProfileSelector);
+
+  // Default our data.
+  let preferences = {};
+  let account = userAccount;
```
Comment (davidjoy) [inline src/sagas/RootSaga.js]: Assume we'll use the current user account data for the page - this is overridden below if we're not looking at the current user's profile.

File: src/selectors/ProfilePageSelector.js [hunk 3]
```diff
@@ -164,12 +165,21 @@ export const handleSaveProfileSelector = createSelector(
...
+  }),
+);
+
 // Reformats the social links in a platform-keyed hash.
 const socialLinksByPlatformSelector = createSelector(
   profileAccountSelector,
   (account) => {
     const linksByPlatform = {};
-    if (account.socialLinks !== undefined) {
+    if (account.socialLinks != null) {
```
Comment (davidjoy) [inline src/selectors/ProfilePageSelector.js]: The `userAccount` reducer has different defaults for socialLinks, so this needed to be made a bit less stringent, along with the one below.  In our profile reducer, we default it to an empty array, whereas in userAccount it's just null.

File: src/sagas/RootSaga.js [hunk 4]
```diff
@@ -38,29 +38,39 @@ import { keepKeys } from '../services/utils';
 
 export function* handleFetchProfile(action) {
   const { username } = action.payload;
-  const currentUsername = yield select(state => state.authentication.username); // eslint-disable-line
+  const { isCurrentUserProfile, userAccount } = yield select(handleFetchProfileSelector);
```
Comment (douglashall) [inline src/sagas/RootSaga.js]: Nit: `isAuthenticatedUserProfile`

File: src/sagas/RootSaga.js [hunk 5]
```diff
@@ -38,29 +38,39 @@ import { keepKeys } from '../services/utils';
...
+
+  // Default our data.
```
Comment (albemarle) [inline src/sagas/RootSaga.js]: I don't think this comment is necessary.

File: src/sagas/RootSaga.js [hunk 6]
```diff
@@ -38,29 +38,39 @@ import { keepKeys } from '../services/utils';
...
+    // Depending on which profile we're loading, we need to make different calls.
     const calls = [
-      call(ProfileApiService.getAccount, username),
+      // We'll always make a call for certificates.
       call(ProfileApiService.getCourseCertificates, username),
     ];
 
-    if (username === currentUsername) {
+    if (isCurrentUserProfile) {
+      // If the profile is for the current user, get their preferences.
```
Comment (albemarle) [inline src/sagas/RootSaga.js]: Same here.  Ideally what's happening should be clear from the code (which it is here); comments should be for when you have to add context to explain _why_ you're doing something.

File: src/selectors/ProfilePageSelector.js (hunk 3 above)
Comment (albemarle) [reply src/selectors/ProfilePageSelector.js]: Should it become a function?  `hasSocialLinks` or similar?

File: src/sagas/RootSaga.js (hunk 6 above)
Comment (davidjoy) [reply src/sagas/RootSaga.js]: I add comments when I feel like the decisions in a piece of code are opaque to someone without any context.  In this function's case, there are specific calls we need to make in different situations, which interacts with how we're setting up the data.  I'm okay with some hand holding if it helps anyone wondering what's going on, personally.  That said, I'm updating the comments to be more specific to the decision process.

File: src/sagas/RootSaga.js (hunk 4 above)
Comment (davidjoy) [reply src/sagas/RootSaga.js]: Doing this. 👍

File: src/selectors/ProfilePageSelector.js (hunk 3 above)
Comment (davidjoy) [reply src/selectors/ProfilePageSelector.js]: Eh... could be, but I don't think it'll see much reuse or anything.

Comment (davidjoy) [reply src/selectors/ProfilePageSelector.js]: Actually, I'm going to change it to use `Array.isArray()` - that'll be more specific to what this is trying to check for.

Comment (albemarle) [reply src/selectors/ProfilePageSelector.js]: I thought you had to change it in two places in this PR-- or is line 195 a different object?

**Pull request merged**